Use AbortController to cancel fetch requests for unmounted components

donnapep opened the pull request:
Fixes the following React warnings and errors when the tests were run:
- `Can't perform a React state update on an unmounted component.`
- `A component is changing an uncontrolled input of type checkbox to be controlled.`

### Changes proposed in this Pull Request
I borrowed and integrated [this solution](https://gist.github.com/kentcdodds/b36572b6e9227207e6c71fd80e63f3b4) to handle the warning. It successfully eliminates the warning and also solves the underlying issue.

### Testing instructions
Ensure that you don't see the above warning / error on the JS tests (they fail for another reason 😅 ).

donnapep committed:
Fix "Can't perform a React state update on an unmounted component" warning

donnapep committed:
Fix "A component is changing an uncontrolled input of type checkbox to be controlled"

gabrielcaires commented on assets/admin/students/student-modal/course-list.js:
```diff
@@ -152,9 +153,12 @@ export const CourseList = ( { searchQuery, onChange } ) => {
-								( { id } ) => id === course.id
-							) }
+							checked={
+								selectedCourses.length > 0 &&
```
@donnapep 
`selectedCourses.current.length`

donnapep committed:
Fix fetch courses tests

gabrielcaires committed:
Manage CourseItem checked state using a internal state.

Previously the code was setting the Checkbox's checked attr directly from the component props, it was triggering the onChange event twice, one when the Checkbox was clicked and the second when the Component received the update checked prop.

gabrielcaires commented on assets/admin/students/student-modal/course-list-fetch.test.js:
```diff
@@ -41,6 +48,7 @@ describe( 'CourseList fetch', () => {
 				expect( httpClient ).toHaveBeenCalledWith( {
 					path: '/wp/v2/courses?per_page=100',
 					method: 'GET',
+					signal: { aborted: false },
```
```suggestion
					signal: new AbortController().signal,
```

gabrielcaires commented on assets/admin/students/student-modal/course-list-fetch.test.js:
```diff
@@ -8,8 +8,14 @@ import { render, waitFor } from '@testing-library/react';
 jest.mock( '../../lib/http-client' );
+jest.mock( './use-abort-controller' );
+
+useAbortController.mockImplementation( () => {
```
You can remove the use-abort-controller mock using  
`signal: new AbortController().signal` as I suggested above.

donnapep replied:
Done in https://github.com/Automattic/sensei/pull/5065/commits/1a0f767ae91aa4cb95791f56c520e580c192ee6b.

gabrielcaires commented on assets/admin/students/student-modal/course-list.js:
```diff
@@ -117,15 +123,14 @@ export const CourseList = ( { searchQuery, onChange } ) => {
-					setIsFetching( false );
-				} )
+				.then( ( result ) => setCourses( result ) )
+				.catch( () => {} )
```
Long comment without an action item: 😂


This line `.catch( () => {} )`  is silencing all errors incoming from the `api-fetch`.
If you remove it and run the tests, you will see a lot of `You are probably offline.`

Investigating a bit more a found that the `api-fetch` version we are using[ hides the abort errors.](https://github.com/WordPress/gutenberg/blob/%40wordpress/api-fetch%403.22.0/packages/api-fetch/src/index.js#L102)

New `api-fetch` versions [are exposing errors incoming from aborted requests.
](https://github.com/WordPress/gutenberg/blob/%40wordpress/api-fetch%403.22.0/packages/api-fetch/src/index.js)

I think, considering our context, we should ignore all abort errors and re-throw the other ones, but considering today we haven't a way to identify them until we update the `api fetch`.

I don't have a suggestion for this case, all scenarios seem not so ideal for me, but I think it is the best we can do.

donnapep replied:
> If you remove it and run the tests, you will see a lot of You are probably offline.

This is the exact reason I put it back in actually. 😆

What do you think about [logging the error to the console](https://github.com/Automattic/sensei/pull/5065/commits/5ec3bb86e086eda348f536944bfdec33fa5ddb10) for now? I'm not sure how useful this will be or what other types of errors are possible besides `fetch_error`, but think it couldn't hurt. It will make the tests a bit noisier though.

gabrielcaires replied:
Yes, I think the log will help us to remember that we need to solve it properly.

I still don't know if something is wrong with our test setup that is causing this error or if the `api-fetch` really should throw an expectation, and is our responsibility to filter this error.

gabrielcaires replied:
According to the official specs, we should just filter the AbortError
https://dom.spec.whatwg.org/#aborting-ongoing-activities-example
So probably we should do it after updating the api-fetch.

gabrielcaires reviewed (dismissed):
Just some comments and suggestions.
I think we should apply the same strategy on the students modal component, but I think it may be in another PR.

I just noticed we are missing the `useAbortController` tests, we need to make sure that we are really aborting the controller on the component unmount.

donnapep committed:
Call AbortController().signal in test data

Co-authored-by: Gabriel Caires <gabriel.caires@automattic.com>

donnapep committed:
Remove use of mock in fetch courses tests

donnapep committed:
Log any course fetch errors to the console

gabrielcaires committed:
Add tests to  use abort controller hook

gabrielcaires committed:
Use abort controller hook on the students modal

m1r0 reviewed (approved):
Looks good to me - I don't see the errors. 🙂
